tt_metal/impl/buffers/buffer.cpp

```diff
@@ -198,89 +227,158 @@ BufferPageMapping generate_buffer_page_mapping(const Buffer &buffer) {
+
+        self->allocation_status_.store(AllocationStatus::ALLOCATED, std::memory_order::relaxed);
+        lock.unlock();
+        self->allocation_cv_.notify_all();
+    });
 }
 
-Buffer &Buffer::operator=(const Buffer &other) {
-    if (this != &other) {
-        this->device_ = other.device_;
-        this->size_ = other.size_;
-        this->page_size_ = other.page_size_;
-        this->buffer_type_ = other.buffer_type_;
-        this->buffer_layout_ = other.buffer_layout_;
-        this->shard_parameters_ = other.shard_parameters_;
-        this->bottom_up_ = other.bottom_up_;
-        this->buffer_page_mapping_ = other.buffer_page_mapping_;
-        this->allocate_ = other.allocate_;
-        if (this->allocate_) {
-            this->allocate();
+void Buffer::deallocate() {
+    if (size_ == 0) {
+        // 0-size buffer, no need to deallocate
+        return;
+    }
+
+    {
+        std::unique_lock lock(allocation_mutex_);
+        auto status = allocation_status_.load(std::memory_order::relaxed);
+        if (status != AllocationStatus::ALLOCATED && status != AllocationStatus::ALLOCATION_REQUESTED) {
+            // Buffer isn't allocated, nothing to be done
+            return;
         }
+        // Overwriting either ALLOCATED or ALLOCATION_REQUESTED with DEALLOCATION_REQUESTED
+        allocation_status_.store(AllocationStatus::DEALLOCATION_REQUESTED, std::memory_order::relaxed);
     }
-    return *this;
-}
-
-Buffer::Buffer(Buffer &&other) :
-    device_(other.device_),
-    size_(other.size_),
-    address_(other.address_),
-    page_size_(other.page_size_),
-    buffer_type_(other.buffer_type_),
-    buffer_layout_(other.buffer_layout_),
-    shard_parameters_(std::move(other.shard_parameters_)),
-    bottom_up_(other.bottom_up_),
-    buffer_page_mapping_(std::move(other.buffer_page_mapping_)),
-    allocate_(other.allocate_) {
-    // Set `other.device_` to be nullptr so destroying other does not deallocate reserved address space that is
-    // transferred to `this`
-    other.device_ = nullptr;
-}
-
-Buffer &Buffer::operator=(Buffer &&other) {
-    if (this != &other) {
-        this->device_ = other.device_;
-        this->size_ = other.size_;
-        this->address_ = other.address_;
-        this->page_size_ = other.page_size_;
-        this->buffer_type_ = other.buffer_type_;
-        this->buffer_layout_ = other.buffer_layout_;
-        this->shard_parameters_ = std::move(other.shard_parameters_);
-        this->bottom_up_ = other.bottom_up_;
-        this->buffer_page_mapping_ = std::move(other.buffer_page_mapping_);
-        this->allocate_ = other.allocate_;
-        // Set `other.device_` to be nullptr so destroying other does not deallocate reserved address space that is
-        // transferred to `this`
-        other.device_ = nullptr;
+
+    device_->push_work([self = weak_self.lock()] {
+        // Because the status is DEALLOCATION_REQUESTED, it won't be changed by anyone else, no need to lock a mutex
```

Comment:
Please consider using the `enable_shared_from_this` base utility class. This will allow you to pass a shared ptr to the current instance without dynamically creating one each time you interface with the worker thread.

I think this will functionally have the same effect and should be safe since the buffer ctor is private.

Reply:
@tt-asaigal  `std::enable_shared_from_this` provides the same functionality as storing a weak_ptr internally, moreover it is actually implemented in this way, so the performance of the two approaches is identical.

The main reason I don't use `std::enable_shared_from_this` here:
- You must inherit from `std::enable_shared_from_this` publicly, which results in an extra `shared_from_this()` public method being available on the buffer, which isn't nice and kinda leaks an implementation detail to the outside.
Also I think it's more straightforward to see what's actually going on here.